horovod/tensorflow/__init__.py

```diff
@@ -66,14 +67,21 @@ def allreduce(tensor, average=None, device_dense='', device_sparse='',
                      using compression.
         op: The reduction operation to combine tensors across different ranks.
             Defaults to Average if None is given.
+        prescale_factor: Multiplicative factor to scale tensor before allreduce.
+        postscale_factor: Multiplicative factor to scale tensor after allreduce.
 
     Returns:
         A tensor of the same shape and type as `tensor`, summed across all
         processes.
     """
     op = handle_average_backwards_compatibility(op, average)
-    # Averaging happens in framework code, so translate that to Sum for the actual call
+
     true_op = Sum if op == Average else op
+    if not rocm_built():
+      if (op == Average):
+        # Averaging happens via postscale_factor.
+        postscale_factor /= size()
```

Comment:
For Elastic mode, having the size defined only during graph construction like this is causing problems when the world size scales up or down.  If possible, it would be good if we could move this into the C++ layer to avoid this problem.

One way we could do this: remove the "true op" concept and just pass the actual op enum to C++.  If it's AVERAGE, we can scale the `postscale_factor` dynamically within the TensorFlow custom op using the updated world size.  Then, we can still pass in SUM to the enqueue function so everything else works the same.

What do you think?

Reply:
I see what you are saying. Perhaps we can move the modification of `postscale_factor` into `EnqueueTensorAllreduce` here, if op is average:
https://github.com/horovod/horovod/blob/b94807d20916a6ac5fd85922e593b2f35f62b4e0/horovod/common/operations.cc#L830-L837

But then this also leads to the question of how to handle the `gradient_predivide_factor` as well. For elastic, are we ok if the `postscale_factor` and `prescale_factor` remain constant w.r.t the world size? In general, the `gradient_predivide_factor` is set based the number of workers.